### PR TITLE
[DependencyInjection] Avoid unnecessary calls to strtolower()

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #14558 |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
View diff as https://github.com/symfony/symfony/pull/14643/files?w=1

File: src/Symfony/Component/DependencyInjection/Container.php [hunk 1]
```diff
@@ -242,17 +244,21 @@ public function set($id, $service, $scope = self::SCOPE_CONTAINER)
...
+        for ($i = 2;;) {
+            if ('service_container' === $id
+                || isset($this->services[$id])
+                || isset($this->aliases[$id])
+                || array_key_exists($id, $this->services)
+                || method_exists($this, 'get'.strtr($id, self::$underscoreMap).'Service')
+            ) {
+                return true;
+            }
+            if (--$i && ($id !== $lcId = strtolower($id))) {
```
Comment (Tobion) [inline src/Symfony/Component/DependencyInjection/Container.php]: the brackets around `$id !== $lcId = strtolower($id)` seem useless

**Comment by Tobion** [comment]
I guess it would be good to have a test that ensures the bug explained by stof in https://github.com/symfony/symfony/pull/14558#issuecomment-100850702 is not triggered.

**Comment by nicolas-grekas** [comment]
@stof @Tobion Any idea which test I should add? I'm not sure to have to right ideas for now.

File: src/Symfony/Component/DependencyInjection/Container.php [hunk 2]
```diff
@@ -76,6 +76,8 @@ class Container implements IntrospectableContainerInterface
     protected $scopeStacks;
     protected $loading = array();
 
+    private static $underscoreMap = array('_' => '', '.' => '_', '\\' => '_');
```
Comment (dosten) [inline src/Symfony/Component/DependencyInjection/Container.php]: Why this is static?

File: src/Symfony/Component/DependencyInjection/Container.php [hunk 3]
```diff
@@ -294,57 +297,60 @@ public function get($id, $invalidBehavior = self::EXCEPTION_ON_INVALID_REFERENCE
...
-            if (self::EXCEPTION_ON_INVALID_REFERENCE === $invalidBehavior) {
-                if (!$id) {
-                    throw new ServiceNotFoundException($id);
-                }
+            if (isset($this->methodMap[$id])) {
+                $method = $this->methodMap[$id];
+            } elseif (--$i && $id !== $lcId = strtolower($id)) {
+                $id = $lcId;
+                continue;
+            } elseif (method_exists($this, $method = 'get'.strtr($id, array('_' => '', '.' => '_', '\\' => '_')).'Service')) {
```
Comment (dosten) [inline src/Symfony/Component/DependencyInjection/Container.php]: Why you don't use `self::$underscoreMap` here?

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Container.php]: I missed it, fixed thanks

File: src/Symfony/Component/DependencyInjection/Container.php (hunk 2 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Container.php]: It's faster and private. But I just made it non static, that won't make a significant difference.

File: src/Symfony/Component/DependencyInjection/Container.php [hunk 4]
```diff
@@ -242,17 +244,21 @@ public function set($id, $service, $scope = self::SCOPE_CONTAINER)
...
-        $id = strtolower($id);
-
-        if ('service_container' === $id) {
-            return true;
+        for ($i = 2;;) {
+            if ('service_container' === $id
+                || isset($this->services[$id])
+                || isset($this->aliases[$id])
+                || array_key_exists($id, $this->services)
+                || method_exists($this, 'get'.strtr($id, $this->underscoreMap).'Service')
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Container.php]: I suggest checking the methodMap before doing the replacements here (as done in ``get``)

**Comment by stof** [comment]
@nicolas-grekas here is what the test should do:

- define a container-scoped service
- dump the container to PHP and continue the work with the dumped container (the bug in the previous PR does not affect the ContainerBuilder as its instantiation logic is different)
- get it from the container (so that it is instantiated)
- get it again from the container with a non-lowercase id, and a different casing than the previous time
- assert that both instances are the same

You can run the test against the code of #14558 to ensure it catches the issue properly.

**Comment by nicolas-grekas** [comment]
@stof thanks, done. I verified that the test fails on #14558 but not here.

**Comment by fabpot** [comment]
:+1:

**Comment by stof** [comment]
:+1:

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Pull request merged**